### ISSUE TITLE
Assign multiple OSC commands to a widget

### ISSUE THREAD

**Issue by git-moss** [jean-emmanuel/open-stage-control#191]
Currently, the OSC command which is assigned to a widget applies to its main content (value) or behaviour (button pushed). To additionally modify another property of a widget complicated workarounds need to be put in place.

Implementation:
- For simple usage the OSC section looks like before but it has a "+" button in it's header
- When clicked, tabs are displayed for each assigned OSC command below the OSC header.
- each tab contains the full set of OSC fields
- On each tab (except the 1st default) you can choose the property to be assigned to the OSC command.
- The name of the tab is the assigned widget property
- At least the following properties would be useful but could also be all properties of a widget: Label, Textcolor, Backgroundcolor, Visibility

**Comment by kdejaeger** [comment]
This would help us a lot because except the value, you want to have the label, color, visibility data, etc of 1 widget also fetched from OSC and consumable into the dynamic properties of the widget. Without having to make an invisible dummy widget somewhere just to be able to consume it's value. It becomes a mess after a while if you want to make soo many things depend on OSC data.

**Comment by jean-emmanuel** [comment]
I heard you guys, I'll experiment on this when I have more time (for now I can only fix little bugs).

**Comment by jean-emmanuel** [comment]
I've been working on an implementation that's doesn't seem to break anything, phew... This needs to be tested and improved: you can use the `OSC{/address, default_value}` in properties to control them  remotely.
- like the `@{}` syntax, it doesn't support nesting 
- like the `@{}` syntax, it can be used in formulas
- under the hood, it creates an osc listener bound to one address that updates the property when it receives a new value
- if the widget has preArgs, those will have to be matched for the listener to process the message

**Comment by kdejaeger** [comment]
Waauw github is amazing detecting references like that. Anyway nice!! I have no time though to check it out completely now. Do we need parentheses around /address?

**Comment by kdejaeger** [comment]
Is there really no way to make this work:?

#{
color = OSC{/track/@{parent.variables.n}/color, 0}
}

**Comment by jean-emmanuel** [comment]
No way that is good enough.

**Comment by kdejaeger** [comment]
Sorry, I don't understand. Is it possible or not?

**Comment by jean-emmanuel** [comment]
No.

**Comment by cyberic99** [comment]
so, this feature allows to get an OSC property from a widget, by chosing it by its address and not by its id?

**Comment by kdejaeger** [comment]
Hmm yes/no, the OSC functionally is not really coming from a widget itself. It's some kind of a function that you can use inside of the dynamic values of a widget. Let's say you have an OSC URL for your widgets css, color, label, etc, you could now fetch this with a OSC URL. Instead of having to create a dummy widget for this just to consume it's value.

**Comment by kdejaeger** [comment]
@jean-emmanuel, is there no workaround you can think of for my issue? I'd like to use cloned strips with OSC, but there's of course always some parent variables that make up the OSC url's.

**Comment by jean-emmanuel** [comment]
@kdejaeger I can't think of anything right now, the solution would be to support nesting, but I havent found a proper way of doing so yet.

**Comment by cyberic99** [comment]
nesting would help in many cases, yes.

@jean-emmanuel are you using regexes to replace things like ``@{}`` with their values?

**Comment by jean-emmanuel** [comment]
@cyberic99 parsing nested structures is not the problem, it's keeping track of all objects and listeners to avoid memory leaks that's a bit tedious in this case.

**Comment by cyberic99** [comment]
I see. Supporting one level of nesting would be sufficient for more cases I think
Maybe one level only is easier to implement?
In addition, it could also be sufficient to resolve the fields one time for all.
For instance if we for a property equals to @{name_@{parent.variables.n}}, you can replace it with name_n once n is known, and that's it.

**Comment by jean-emmanuel** [comment]
I've implemented a limited nesting feature : 
- `@{}` calls can be nested, but nested calls are not dynamic (resolved only once at widget's initialization).
- `OSC{}` and `#{}` calls can contain full featured `@{}` calls

**Comment by kdejaeger** [comment]
Sounds nice. This at least fixes the 'strip counter' issue for things like #{ color = OSC{/track/@{parent.variables.n}/color, 0} }. And I could put a dynamic label at push buttons inside a strip.

**Comment by cyberic99** [comment]
@jean-emmanuel what do you call 'full featured'?
you mean fully dynamic?

**Comment by jean-emmanuel** [comment]
@cyberic99 I mean dynamic for the first level of `@{}` calls.
@kdejaeger I could use some test feedback before releasing though, just sayin' ,)

**Comment by jean-emmanuel** [comment]
Nevermind I'll just release it as a beta and travis will do the builds.

**Comment by jean-emmanuel** [comment]
https://github.com/jean-emmanuel/open-stage-control/releases/tag/v0.29.1-beta

**Comment by kdejaeger** [comment]
This doesn't seem to work yet:

#{
css = "color:  OSC{/track/@{parent.variables.n}/color}"
}

I try this on a push button

**Comment by jean-emmanuel** [comment]
It does work here once i send a message to `/track/@{parent.variables.n}/color}`. Note that changing the css color property is different from changing the widget's color property: the first one affects the default text color, the latter affects the custom accent color (which sets the css variable `--color-custom` under the hood).

**Comment by jean-emmanuel** [comment]
> v0.29.1
Feel free to open a new issue if something goes wrong.

**Comment by kdejaeger** [comment]
I got it too work now too. But it took me a while, I think suddenly when I opened with F12 the google chrome inpector, it started working ... caching issue? Don't know for sure.
